grip_api/src/grip_api/abstract_widgets/base_text_editor.py

```diff
@@ -0,0 +1,195 @@
+#!/usr/bin/env python3
+
+# Copyright 2023 Shadow Robot Company Ltd.
+#
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU General Public License as published by the Free
+# Software Foundation version 2 of the License.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program. If not, see <http://www.gnu.org/licenses/>.
+
+import abc
+from typing import Union
+from PyQt5 import Qsci
+from PyQt5.QtGui import QColor
+from PyQt5.QtCore import pyqtSignal, QTimer, Qt
+
+
+class MetaTextEditor(type(Qsci.QsciScintilla), abc.ABCMeta):
+    """
+        Meta class that allows for creating abstract base classes for text editors, inherited from Scintilla
+    """
+    # When the text editor is empty (i.e. no new page or anything), define the background color as grey
+    EMPTY_COLOR = QColor("#cccccc")
+    # Amount of time (in ms) to wait since the last text edition to trigger parsing and validity checks
+    TIME_BEFORE_PARSING = 600
+    # Color of the background set to the lines that are wrongly formatted
+    WRONG_FORMAT_COLOR = QColor("#40FF0000")
+
+
+class BaseTextEditor(Qsci.QsciScintilla, abc.ABC, metaclass=MetaTextEditor):
+    """
+        QScintilla-based abstract text editor
+    """
+    # Signal sent when the parsed content changes
+    contentIsModified = pyqtSignal(bool)
+
+    def __init__(self) -> None:
+        """
+            Initialize the class by setting up the editor
+        """
+        super().__init__()
+        self.text_lexer = None
```

Comment:
Google...

Reply:
> Should we add a getter/setter for this?

Ok, so it's my bad since I did not format this right (thanks for pointing it out) but no setter or getter needs to be done here since it's not supposed to be accessible / changeable once it is set

Reply:
Updated here: [eb3d0e8](https://github.com/shadow-robot/sr_grip/pull/17/commits/eb3d0e8bb7238aad5be85a7de4c132fc63b0270a)